### ISSUE TITLE
Get MacroPy working on Python 3.4

### ISSUE THREAD

**Issue by lihaoyi** [lihaoyi/macropy#32]
I have no idea how hard this will be, but MacroPy does not have many dependencies on the underlying runtime except for PEP302 and the `ast` library.

Some of the ASTs look slightly different (e.g. functions can't have nested parameter lists, no `print` statement) and we may have to remove any `print` statements from the implementation code, but I don't imagine it will be very difficult or require big changes.

**Comment by ceridwen** [comment]
What still needs to be done for this branch?  I'd like to explore MacroPy on 3.4.

**Comment by lihaoyi** [comment]
Take the python3 branch and see whether your can build it and what breaks ^_^

**Comment by shawndfisher** [comment]
From the first example:

Python 3.4.2 (default, Jan 12 2015, 11:38:40) 
[GCC 4.9.2 20141224 (prerelease)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import macropy.console
0=[]=====> MacroPy Enabled <=====[]=0
>>> from macropy.case_classes import macros, case
>>> @case
... class Point(x, y): pass
... 
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/sfisher/dev/macropy/macropy/core/failure.py", line 17, in raise_error
    raise ex
macropy.core.failure.MacroExpansionError: arguments constructor takes either 0 or 6 positional arguments
. . .

I'll look at this a bit more when I get home.

**Comment by ceridwen** [comment]
I've been working on this.  Before I start making pull requests, would you prefer if I lump everything together into one request or break it up into smaller logical units?

**Comment by lihaoyi** [comment]
Just lump it all together ^_^

**Comment by ceridwen** [comment]
Does anyone care about Python 3.3 and earlier compatibility?  I don't, and I can't imagine anyone else wants to use this library on Python 3 and hasn't updated to 3.4, but I could be wrong so I thought I'd ask first.

**Comment by ChadSki** [comment]
When I first commented in 2013 I cared, but no longer! =)

**Comment by ceridwen** [comment]
I'm almost finished with refactoring all the references to the ast module to use `ast.foo` rather than `foo`, in preparation for moving to some method of handling the differences between ASTs in 2 and 3.  If I'm going to write classes to wrap the _ast classes and hide the differences between the two versions, it's only some more work to write something like the module described in #9.   However, my first inclination is to use https://bitbucket.org/logilab/astroid (the successor to astng) instead.  It already supports 2.7 and 3.3+ in the same code base and is vastly better tested than anything I could ever write.  It would introduce two additional dependencies, astroid itself and the logilab-common package---six is already necessary for the python3 branch.  Does anyone have an opinion one direction or another?  I haven't examined astroid in detail yet, so comments from anyone who has would be appreciated.

As for the compatibility question I posed in my last comment, astroid would guarantee 3.3+ compatibility.  The only reason I care about compatibility with any version of 3 less than 3.3 is because PyPy 3 is still on 3.2, but if the only reason that astroid doesn't support 3.2 is because of the u'' issue (the usual reason), it should run on PyPy 3 anyways.

**Comment by ceridwen** [comment]
I hadn't noticed when I wrote my last comment, but one possible issue with astroid: it's licensed under the GPL.  I don't know if that would forbid use in Macropy, which is MIT licensed.  If anyone knows for sure, that would be useful.

**Comment by paulie-g** [comment]
It's not GPL licensed, the license for astroid is LGPL. Without going into detail, you can import it and list it as a dependency for the purposes of automatic installation (eg with pip) without any issues.

**Comment by ceridwen** [comment]
Good point, I somehow must have mistaken its license for pylint's or one of the other related projects distributed by Logilab.  Since that shouldn't affect MacroPy's licensing, any objects to my using astroid to get 2/3 compatibility for the ASTs?  I've been busy lately so I haven't had the chance to do any real work on Macropy, but after next week I should have some time again.

**Comment by paulie-g** [comment]
I haven't contributed to macropy yet, so my +1 in favour of using an ast abstraction layer such as astroid may not be worth much, but there it is nonetheless. Keeping up with ast changes across python versions is a maintenance burden that would make a lot of sense to shift onto someone else, imho. Since astroid is central to pylint, there's an excellent chance that it will continue to be actively maintained, so it's a good choice from that point of view. From a technical point of view, ideally we'd need it to abstract the ast sufficiently for macropy to work across multiple minor cpython versions. Does it do that?

**Comment by lihaoyi** [comment]
I'm not going to contribute much, but as https://github.com/lihaoyi/macropy/issues/9 shows I support having a better, more friendly AST to use inside macros ^_^

**Comment by paulie-g** [comment]
Btw, does astroid have a semblance of documentation anywhere or am I reduced to just reading the code?

**Comment by ChadSki** [comment]
The best I found is at [readthedocs](http://astroid.readthedocs.org/en/latest/) but there are more TODO statements than actual content.

**Comment by thedrow** [comment]
@lihaoyi Maybe it's a good time to find a new maintainer for this project since it has a lot of potential.

**Comment by paulie-g** [comment]
I agree that MacroPy has a lot of potential. With that said, even if a new maintainer were necessary and @lihaoyi wanted one, new maintainers are best chosen from a set of people who have made substantial code contributions already. It looks like this set currently = @lihaoyi ;) He seems active enough that getting pull requests (once there are some) reviewed and merged shouldn't be a problem.

In any case, that discussion is probably best handled in a separate issue since it's orthogonal to this one.

**Comment by lihaoyi** [comment]
@thedrow as the person who thinks the project has potential finding a new maintainer is your job, not mine ^_^

**Comment by ds2643** [comment]
What's the status on this project?  Is Python 3 currently supported? I'm happy to volunteer to help with the effort if needed.

**Comment by scherrey** [comment]
Also interested in the status of python 3 support and the future of this promising library.

**Comment by raingloom** [comment]
I would also love to see this lib on Python 3

**Comment by azazel75** [comment]
I'm working on the port on my master branch, where I have merged @lihaoyi 's master and the python3 branch and also merged in contribution by some other developer that started working on this long before me but left the ranch in an incomplete, non functional state.

In my master branch there's no python 2 compatibility code ( there is still some left but i'm going to drop it ) and it should work on python 3.5+. All of the core tests pass and mosto of the other as well, there is just someting wrong with the PEG tests

**Comment by lihaoyi** [comment]
@azazel75 :+1 :+1 :+1

**Comment by azazel75** [comment]
all the tests pass, i've also updated pinq to work... Now there is the need to update the code here and there to take into account the "async"  variant of some statement and expressions, like async for, async with, async comprehensions

**Comment by thedrow** [comment]
@azazel75 Can you create a PR to this repo? I've added the Travis CI script which we should extend for Python 3 as well.

**Comment by lihaoyi** [comment]
@azazel75 @thedrow if you guys want to do serious work on this, I can give both of you access to this repo. You could also fork it, though working on the main repo would make sure your works gets the attention of anyone who's interested in this project, which would be more difficult on a fork.

I could probably give you publishing access to PyPI too, except that it's been so long ago I don't remember how PyPI works or what (if any) my credentials are...

**Comment by azazel75** [comment]
Well, it's still unfinished and it's my take on macropy. I mean that as i have no time to debug it on non current pythons, I dropped all the Python 2 and Python 3.4- (maybe also with 3.4 included, i have to check out this to be sure) code. This is mostly beacuse function's arguments have changed very much in their format, and there was code to cast them as they were defined in Py2 and this was causing some problems. While using tools like astroid maybe a solution, it is really poorly documented and at the same time macropy seemed a bit... abandoned. So I draw a line and started fresh with Python 3.5 and 3.6 as targets. This was a decision I made because i needed the tool and it may not be the general consensus on what macropy should target as interpreters. So i'm not so fond that a PR is the right thing to do... in the meantime you can always add a remote to my repo with `git remote add azazels git@github.com:azazel75/macropy.git`, but I'm sure you know already.

**Comment by Daenyth** [comment]
I'd also be interested in helping with python3 support and general development. I feel like this could make my work stuff easier, and I'm quite familiar with 2+3 compatibility issues and testing approaches.

**Comment by lihaoyi** [comment]
Going forward on latest version of Python 3.6 seems totally reasonable to me. Python3 is the future.

**Comment by Daenyth** [comment]
I'd be happy to help either with 2.7/3.5+ support or only 3.6+ support. There's merits for both. 3.6 brings some real benefits with f-strings, async support, type annotations / mypy, and typing.NamedTuple class syntax

**Comment by azazel75** [comment]
@lihaoyi as you are pleased, I've started using it in my python to js transpiler and want to use it to have a more layered approach on transpiling, with the help of type inference (and typescript definitions), so i'll give it some love ;-)

**Comment by thedrow** [comment]
Last time I tried introducing astroid to MacroPy I bumped into this https://github.com/PyCQA/astroid/issues/94
@azazel75 I do think that having Python 2.7 support is beneficial for PyPy users especially. I'd support a Python 3 only codebase if PyPy3 was out of beta. Currently MacroPy causes PyPy's loop unrolling optimization to segfault. I've filed https://bitbucket.org/pypy/pypy/issues/2576/core-dumped-when-running-macropy-tests-on
Other than that PyPy works just the same as CPython 2.7.
Just let me know how I can help.

**Comment by azazel75** [comment]
@thedrow I think that we should reach the goal to run with a current python and fully support it. Then we can decide where to go from there. About astroid, it frightens me to se how many outstanding bugs there are and I think that if the goals are to harmonize Py2 and Py3 calling conventions in a way that informations present in Py3 ast aren't lost or degraded and to have a somewhat more friendly ast base that plays well and is more hackable is no such big deal, but it's a "not now" for me, the port has to be completed first, together with doc refreshing to let interested users know what's going on.

About support for Py2.7, well, it's dead for me. I don't use it anymore and I'm not going to maintain support for it. If you really use it and you want to spend hours debugging on it the I'll leave it to you ;-)

**Comment by azazel75** [comment]
There is good documentation around about the differences on each version's ast

**Comment by thedrow** [comment]
Sounds good to me. Let's start with Python 3 and optionally the PyPy3 beta (we'll allow it to fail in the build matrix) and work our way from there.

**Comment by azazel75** [comment]
So if you want to help, start from my branch, try it out, add support for async stuff, and so on, then we will wait for @lihaoyi  decision and we'll decide .... where to merge ;-) About the doc (the readme), i'm in the process of converting it to rst, because I'm more familiar with it and also think that there is too much in that readme that can generate some confusion... reading an  intro that never ends :-) The format change could come in handy for a later split up in an overview+intro and and a sphinx doc

**Comment by azazel75** [comment]
ah, there is also JoinedStr from 3.6.... that the PyPy guys backported to their 3.5 version of the interpreter... wtf, anyway... its doc is here http://greentreesnakes.readthedocs.io/en/latest/nodes.html?#JoinedStr

**Comment by ceridwen** [comment]
I've been really busy in real life, so I haven't been able to take updating Macropy to Python 3 as far as I'd like.  I have some work in https://github.com/ceridwen/macropy, but there's still quite a ways to go.  I would encourage anyone hoping to port it to look at it, since I fixed some annoying bugs.

I've also done a lot of work at improving astroid's AST for use with Macropy, but there's still more to be done and realistically I don't know when I'm going to get to it.  I just started a new job, so my free time for open source is quite limited right now, and the other major astroid maintainer is also busy.  I do think that using astroid would have advantages.  Python 2.7's EOL isn't until 2020, and there's enormous amounts of legacy code that still exists.  As a library author, I still feel compelled to write Python 2/3 compatible code.  That said, there are also some difficulties.  Here are the biggest ones I've run into:

* astroid's nodes have pointers to their parents as well as their children.  This used to be much worse to work with, I've improved it by creating real `__init__()` and `postinit()` methods for every node class, but initialization still takes two steps, which will require reworking all the places where nodes are built in Macropy.  I have some future plans to turn this into a singly-linked tree, but I haven't finished them because they require extensive refactoring.

* astroid uses the standard library AST parser and then transforms the tree to convert it into an astroid AST.  Custom parsers aren't possible at the moment.

* astroid's static analysis is deeply entwined with the rest of the AST code in some ways that are particular problems for Macropy.  The worst is that astroid stores information about run-time properties like local variables on the AST nodes themselves, and it adds these attributes to the nodes during tree creation using the static analysis code.  If you create an astroid tree without creating these attributes, the static analysis code could misfire badly if it's called on *any* tree containing these new nodes.  However, I don't know what will happen if you tried to call the attribute-assignment on the root of a tree with attributes already assigned to it.  astroid actually does some Macropy-style tree mutation to handle some libraries that do things beyond the capability of its static analysis to analyze and has run into some problems with doing it in a way that doesn't break the static analysis.  The right solution to this is to not use mutable objects on the nodes themselves for the static analysis, but that will take another large amount of work.  Macropy does some limited static analysis itself, which could be replaced with astroid static analysis, but only if these other issues are solved.

* astroid is almost entirely undocumented.

Aside from getting Python 2/3 compatibility, there are also some other benefits:

* Macropy currently has its own methods for converting ASTs into source code, but astroid provides that functionality.  I found a number of bugs in Macropy's implementation in my testing, and there are probably others that I haven't found yet.  astroid's implementation is better-tested and has fewer bugs.

* The Python AST is a moving target, it's changed in every recent Python3 release.  Keeping up with Python 3 support in astroid and Macropy involves duplication of effort.

If anyone wants further advice on this or to ask questions about astroid, I can answer them.

**Comment by azazel75** [comment]
Thanks  for your work @ceridwen,  I based my branch on yours so nothing has been lost.

You say that you found bugs in macropy ast conversion code... do you remember them?
Ther is any alternative to astroid? (Already done and maintained, i mean)

**Comment by ceridwen** [comment]
It's been long enough that I don't, unfortunately.  I'll look through my notes and see if I wrote anything down about them.

There isn't really an alternative to astroid, unfortunately.  There are some other libraries that do static analysis in service of refactoring or code completion with their own tree classes, but those are (AFAIK) concrete syntax trees that are even more inappropriate for Macropy than they are for Pylint.  There's been some discussion about collaborating to improve Python's collection of tree libraries that's never gotten off the ground because of the different needs of static analysis, code completion, and refactoring tools and because of lack of time on the part of the authors.

**Comment by lihaoyi** [comment]
If you really want to make MacroPy work cross-{2.7,3.6}, doesn't [MyPy](https://github.com/python/mypy) have cross-version parsers and a cross-version AST module that you could C&P into MacroPy? My understanding is that MyPy works for both 2.7 and 3.6, and I imagine their "main" type-checking logic isn't implemented twice.

**Comment by azazel75** [comment]
I think they use 'typed-ast', or was it named something else?

**Comment by ceridwen** [comment]
Yeah, mypy now uses two parsers that are part of the typed-ast package.  As far as I know, they're modifications of the CPython 2 and 3 parsers.  Their Github page says they have experimental support for converting Python 2 ASTs into 3 ASTs, but I don't think that helps for Macropy.

**Comment by thedrow** [comment]
@azazel75 Can you enable Travis on your repo so PRs will be easily tested?

**Comment by thedrow** [comment]
I run all the tests and everything passes on 3.6. What is missing?

**Comment by azazel75** [comment]
@thedrow, Travis it's enabled now.

What it lacks mostly is a full coverage of the Python 3.5-3.6 syntax, in https://github.com/azazel75/macropy/blob/master/macropy/core/__init__.py we miss handling of async ( for, with, async compehension), await, and the already mentioned ast.JoinedStr node  ( the node for `f"Hello from {planet}"`), there is some marked TODO: that needs to be completed :-/.  The fact is that the testsuite  tests only a subset of the nodes available in a language release. I was thinking about a way to test this,  about testing what's handled by macropy at least in terms of parsing/repr capability and know what's missing automatically,  I have similar needs in https://github.com/azazel75/metapensiero.pj ( see https://github.com/azazel75/metapensiero.pj/blob/master/tests/test_ast/test_ast_dump/fstring_format_3.6.0.py and https://github.com/azazel75/metapensiero.pj/blob/master/tests/test_ast/test_ast_dump/fstring_format_3.6.0.ast  for example) was thinking about sharing some or part of the testuite and the overall value/cost of such a test...

**Comment by thedrow** [comment]
Let's continue this discussion in email/chat somewhere.

**Comment by fferri** [comment]
Try https://gitter.im

**Comment by azazel75** [comment]
I'm azazel on freenode's IRC. I'm always connected during work days, i'll join a channel named #macropy laterali today or tomorrow 

Il 17 giugno 2017 10:57:30 CEST, Omer Katz <notifications@github.com> ha scritto:
>I run all the tests and everything passes on 3.6. What is missing?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/lihaoyi/macropy/issues/32#issuecomment-309202997

-- 
Inviato dal mio dispositivo Android con K-9 Mail. Perdonate la brevità.

**Comment by azazel75** [comment]
If you prefer anything else, just let me know.

**Comment by lihaoyi** [comment]
I went and opened a gitter chat room for lihaoyi/macropy:

- https://gitter.im/lihaoyi/macropy

You guys can use it, or not, up to you

**Comment by azazel75** [comment]
I've completed the port to Python 3.4+ and I've revamped the docs... I've already published my fork under the name macropy3 because I needed it in another project and I needed it be published.  I can submit a pull request I you guys think it's better (or you can just merge it) let me know... the docs are here http://macropy3.readthedocs.io/en/latest/ and they are still in the works ;-) let me know what you think

**Comment by azazel75** [comment]
I'm closing this as Python 3.4+ support has landed
